draft/0001-organization.md

```diff
@@ -0,0 +1,273 @@
+
+### 4.5. Assign roles to organization members
+
+Before assigning roles to an identity, the identity MUST have a membership with the organization, making them an organization member. An organization member can be assigned multiple roles within the organization, and each role held by the member must be predefined in the organization template. The complete set of permissions granted to an organization member is the union of all permissions granted by their roles.
+
+### 4.6. Role-based access control
+
+To authorize an actor (identity) to perform a specific action within an organization context, the actor MUST be assigned at least one role that contains the desired permission (scope).
+
+If the actor does not specify the organization context, the authorization MUST fail.
+
+It's important to note that organization roles are not considered in authorization. The actor is authorized to perform the action based solely on the organization permission, regardless of their organization role.
+
+### 4.7. Relationship to existing specifications
+
+Organization permissions (scopes) are distinct from the `scope` in OpenID Connect and [Resource Indicators](https://www.rfc-editor.org/rfc/rfc8707.html#name-access-token-request). While these existing specifications share terminology like "scope" and "role" with organization, they are designed to address different authentication and authorization challenges. Organization scopes and roles should be treated separately.
+
+## 5. Authentication
+
+In OpenID Connect, the [ID Token](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) is a JWT token containing user identity information. In addition to the [Standard Claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims), we introduce new claims to the ID token to support organization.
+
+### 5.1 Organization claims
+
+| Name               | Type                  | Description                                                                                                                                                                                          |
+| ------------------ | --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| organization_ids   | array of strings      | The unique organization identifiers that the user has a membership with.                                                                                                                             |
+| organization_roles | array of JSON objects | An array of organization roles that the user has. Each object contains the following fields: <ul><li>`organization_id`: The unique organization identifier as a string.</li><li>`roles`: The role names granted to the user within the organization.</li></ul> |
```

Comment:
To simplify the data structure. It is acceptable to make the `orginazation_roles` as an array of strings as well. 
orginationRoleName could have a standard format sth like `${orginzationId}:${roleName}`?

And also maybe a single claim field called `organizations`:

```
{
 id: string;
 roles: Role[]
}
```

Reply:
yeah i thought about it, but it's not developer-friendly and it needs extra parsing in the client side. i know it's ok, i just don't understand the drawback of following the standard json format (i mean for `scope` in OAuth), since it brings more troubles than the benefit (size reduction), and the size of tokens should not be a consideration in most reasonable cases nowadays.

Reply:
the issue with a single `organizations` field: a user may have no role in some orgs.

Reply:
use
```
{ id: string; roles: Role[] }
```